File: src/main/java/main/commands/slashcommands/commands/dev/ResetPremiumFeaturesCommand.java
package main.commands.slashcommands.commands.dev;

import main.main.Robertify;
import main.utils.RobertifyEmbedUtils;
import main.utils.apis.robertify.models.RobertifyPremium;
import main.utils.component.interactions.AbstractSlashCommand;
import main.utils.locale.RobertifyLocaleMessage;
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import net.dv8tion.jda.api.interactions.commands.OptionType;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;

public class ResetPremiumFeaturesCommand extends AbstractSlashCommand {
    private final Logger logger = LoggerFactory.getLogger(ResetPremiumFeaturesCommand.class);

    @Override
    protected void buildCommand() {
        setCommand(
                getBuilder()
                        .setName("resetpremiumfeatures")
                        .setDescription("Reset premium features for a specific guild or all guilds")
                        .addSubCommands(
                                SubCommand.of(
                                        "all",
                                        "Reset premium features for all guilds"
                                ),
                                SubCommand.of(
                                        "guild",
                                        "Reset premium features for a specific guild",
                                        List.of(
                                                CommandOption.of(
                                                        OptionType.NUMBER,
                                                        "guildid",
                                                        "The ID of the guild",
                                                        true
                                                )
                                        )
                                )
                        )
                        .setDevCommand()
                        .build()
        );
    }

    @Override
    public String getHelp() {
        return null;
    }

    @Override
    public void onSlashCommandInteraction(@NotNull SlashCommandInteractionEvent event) {
        if (!devCheck(event)) return;

        final var cmdGuild = event.getGuild();
        event.deferReply().queue();

        try {
            switch (event.getSubcommandName()) {
                case "all" -> {
                    Robertify.getShardManager()
                            .getGuildCache()
                            .forEach(g -> {
                                RobertifyPremium.resetPremiumFeatures(g);
                                logger.info("Reset all premium features for {}", g.getName());
                            });
                }
                case "guild" -> {
                    final var guildID = event.getOption("guildid").getAsLong();
                    final var guild = Robertify.getShardManager().getGuildById(guildID);
                    if (guild == null) {
                        event.getHook().sendMessageEmbeds(RobertifyEmbedUtils.embedMessage(cmdGuild, "There was no such guild with the ID: " + guildID).build())
                                .setEphemeral(true)
                                .queue();
                        return;
                    }

                    RobertifyPremium.resetPremiumFeatures(guild);
                    logger.info("Reset all premium features for {}", guild.getName());
                    event.getHook().sendMessageEmbeds(RobertifyEmbedUtils.embedMessage(cmdGuild, "Successfully reset all premium features for " + guild.getName()).build())
                            .setEphemeral(true)
                            .queue();
                }
            }
        } catch (Exception e) {
            logger.error("Unexpected error", e);
            event.getHook().sendMessageEmbeds(RobertifyEmbedUtils.embedMessage(
                    cmdGuild, RobertifyLocaleMessage.GeneralMessages.UNEXPECTED_ERROR
            ).build())
                    .setEphemeral(true)
                    .queue();
        }
    }
}

File: src/main/kotlin/main/commands/slashcommands/dev/ResetPremiumFeaturesCommand.kt
package main.commands.slashcommands.dev

import dev.minn.jda.ktx.util.SLF4J
import main.commands.slashcommands.SlashCommandManager.getRequiredOption
import main.utils.RobertifyEmbedUtils.Companion.sendEmbed
import main.utils.api.robertify.models.RobertifyPremium
import main.utils.component.interactions.slashcommand.AbstractSlashCommand
import main.utils.component.interactions.slashcommand.models.SlashCommand
import main.utils.component.interactions.slashcommand.models.CommandOption
import main.utils.component.interactions.slashcommand.models.SubCommand
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent

class ResetPremiumFeaturesCommand : AbstractSlashCommand(
    SlashCommand(
        name = "resetpremiumfeatures",
        description = "Reset premium features for a specific guild or all guilds.",
        developerOnly = true,
        subcommands = listOf(
            SubCommand(
                name = "all",
                description = "Reset premium features for all guilds."
            ),
            SubCommand(
                name = "guild",
                description = "Reset premium features for a specific guild.",
                options = listOf(
                    CommandOption(
                        name = "id",
                        description = "The ID of the guild"
                    )
                )
            )
        )
    )
) {

    companion object {
        private val logger by SLF4J
    }

    override fun handle(event: SlashCommandInteractionEvent) {
        event.deferReply(true).queue()

        when (event.subcommandName) {
            "all" -> handleAll(event)
            "guild" -> handleGuild(event)
        }
    }

    private fun handleAll(event: SlashCommandInteractionEvent) {
        event.jda.shardManager!!.guildCache.forEach { guild ->
            RobertifyPremium.resetPremiumFeatures(guild)
            logger.info("Reset all premium features for ${guild.name}")
        }
        event.hook.sendEmbed(event.guild, "Reset premium features for all guilds!")
            .queue()
    }

    private fun handleGuild(event: SlashCommandInteractionEvent) {
        val guildId = event.getRequiredOption("id").asString
        val eventGuild = event.guild!!
        val guild = event.jda.shardManager!!.getGuildById(guildId)
            ?: return event.hook.sendEmbed(eventGuild, "There was no such guild with the ID: $guildId")
                .queue()

        RobertifyPremium.resetPremiumFeatures(guild)
        logger.info("Reset all premium features for ${guild.name}")
        event.hook.sendEmbed(eventGuild, "Successfully reset all premium features for ${guild.name}")
            .queue()
    }

}